Add extension guides to terraform provider (#590)
nitrictech/docs: src/pages/guides/terraform/s3-encryption.mdx

  advanced encryption options in Amazon S3 may incur costs. Amazon S3 offers
  pricing details on their [official pricing
  page](https://aws.amazon.com/s3/pricing/).
</Note>

## What we'll be doing

1. Review our existing module
2. Configure Server-side encryption (SSE)

## Review our existing module

The first step is clone the [Nitric repository](https://github.com/nitrictech/nitric) and examine how the [Nitric Terrraform provider provisions an S3](https://github.com/nitrictech/nitric/tree/main/cloud/aws/deploytf/.nitric/modules/bucket) bucket.

```bash
git clone https://github.com/nitrictech/nitric
cd nitric
```

The AWS S3 module in the default Terraform provider performs the following tasks:

1. Creates a unique ID for the S3 bucket to ensure unique naming.
2. Provisions an S3 bucket with a unique name using the generated ID.
3. Tags the bucket for identification.
4. Grants S3 permission to invoke specified Lambda functions.
5. Configures S3 bucket notifications to trigger Lambda functions based on specified events using dynamic blocks.

To begin our customization, we will start adding configuration to this module.

## Add Server Side Encryption (SSE)

> Introduce the following configuration into the aws/deploytf/.nitric/modules/bucket/main.tf:

Reviewer: ```suggestion
Introduce the following configuration into aws/deploytf/.nitric/modules/bucket/main.tf:
```

> ```hcl

Reviewer: Could add a tag:

```suggestion
```hcl {{ tag: 'aws/deploytf/.nitric/modules/bucket/main.tf' }}
```

# AWS S3 bucket server-side encryption configuration
resource "aws_s3_bucket_server_side_encryption_configuration" "bucket_encryption" {
  bucket = aws_s3_bucket.bucket.bucket

  rule {
    apply_server_side_encryption_by_default {
      sse_algorithm = "AES256"
    }
  }
}
```

<Note>
  Full documentation can be found on the [Terraform
  registry](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_server_side_encryption_configuration).
</Note>

## Building and using your updated provider

The Nitric project includes a make file that will build and install your provider as nitric/awstf@0.0.1 by default.

> Run `make install` from the provider's root directory - `nitric/cloud/aws`

Reviewer: Just a note: this won't currently work on windows due to the .exe but I will do a PR that fixes this.

The provider can then be used directly in your project's stack file as follows.

```yaml
# The nitric provider to use
provider: nitric/awstf@0.0.1

# The target aws region to deploy to
region: us-east-2
```

<Note>
  If you don't have a stack file use the `nitric stack new` to create one.
</Note>

You'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:

```
preview:
  - beta-providers
```

You can generate the Terraform code by running the following command:

```bash
nitric up
```

To deploy the application using Terraform, you can navigate into your Terraform stack directory and use the standard Terraform commands:

```bash
terraform init
terraform plan
terraform apply
```

> You can examine your configuration by logging into the AWS console.

Reviewer: Is there a link to the aws console we can provide (probably on us-east-2)?

> You can examine your configuration by logging into the AWS console.

Reviewer 2: ```suggestion
Finally, log into the AWS console to verify the encryption configuration was applied.
```

export const description =

>   'Extend the Nitric S3 Terraform provider to replicate your S3 bucket into multiple regions'

Reviewer 3: ```suggestion
  'Extend the Nitric AWS S3 Terraform provider to add server-side encryption'
```

> # Replicate S3 buckets into another region

Reviewer 3: ```suggestion
# Add server-side encryption
```

<Note>
  Server-side encryption with Amazon S3 managed keys (SSE-S3) is free. Other
  encryption options may incur costs. Amazon offers pricing details on their
  [pricing page](https://aws.amazon.com/s3/pricing/).
</Note>

## What we'll be doing

1. Review the existing module
2. Configure server-side encryption (SSE)

## Review the existing module

Start by cloning the [Nitric repository](https://github.com/nitrictech/nitric), then examine how the Terrraform provider [provisions an S3 bucket](https://github.com/nitrictech/nitric/tree/main/cloud/aws/deploytf/.nitric/modules/bucket).

```bash
git clone https://github.com/nitrictech/nitric
cd nitric
```

The AWS S3 module in the default Terraform provider performs the following tasks:

1. Creates a unique ID for the S3 bucket to ensure unique naming.
2. Provisions an S3 bucket with a unique name using the generated ID.
3. Tags the bucket for identification.
4. Grants S3 permission to invoke specified Lambda functions.
5. Configures S3 bucket notifications to trigger Lambda functions based on specified events using dynamic blocks.

To begin our customization, we will start adding configuration to this module.

> ## Add Server Side Encryption (SSE)

Reviewer 3: ```suggestion
## Add server-side encryption (SSE)
```

Add the following at the end of `aws/deploytf/.nitric/modules/bucket/main.tf`:

```hcl {{ tag: 'aws/deploytf/.nitric/modules/bucket/main.tf' }}
# AWS S3 bucket server-side encryption configuration
resource "aws_s3_bucket_server_side_encryption_configuration" "bucket_encryption" {
  bucket = aws_s3_bucket.bucket.bucket

  rule {
    apply_server_side_encryption_by_default {
      sse_algorithm = "AES256"
    }
  }
}
```

<Note>
  Full documentation can be found on the [Terraform
  registry](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_server_side_encryption_configuration).
</Note>

## Build and use your modified provider

The Nitric project includes a make file that will build and install your provider as `nitric/awstf@0.0.1` by default.

Navigate to `nitric/cloud/aws` and run `make install` to build and install the modified provider binary.

```bash
cd nitric/cloud/aws

make install
```

The provider can then be used directly in your project's stack file as follows.

```yaml
# The nitric provider to use
provider: nitric/awstf@0.0.1

# The target aws region to deploy to
region: us-east-2
```

<Note>

>   If you don't have a stack file use the `nitric stack new` to create one.

Reviewer 3: ```suggestion
  If you don't have a stack file use `nitric stack new` to create one.
```

</Note>

> You'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:

Reviewer 3: ```suggestion
Because the [Terraform providers](https://nitric.io/docs/reference/providers/terraform) are in preview, you'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:
```

> Because the [Terraform providers](https://nitric.io/docs/reference/providers/terraform) are in preview, you'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:

Reviewer: ```suggestion
Because the [Terraform providers](/reference/providers/terraform) are in preview, you'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:
```
